### DISCUSSION TITLE
Error on PySpark deduping example: no jars directory in splink

### DISCUSSION THREAD

**Discussion by mmagoffin-sd** [moj-analytical-services/splink#864] [Q&A] [answered]
I just moved from using splink with DuckDB to using it with Spark. I tried to follow the first notebook cell, but I get the following error:

<img width="1314" alt="Screen Shot 2022-10-25 at 3 08 31 PM" src="https://user-images.githubusercontent.com/85239124/197861049-3b5f155c-4674-4cc4-821f-e470ab05aaf0.png">
<img width="1307" alt="Screen Shot 2022-10-25 at 3 08 41 PM" src="https://user-images.githubusercontent.com/85239124/197861070-f5b7217b-c14f-4c69-81aa-0bf52677ecbb.png">
<img width="767" alt="Screen Shot 2022-10-25 at 3 08 47 PM" src="https://user-images.githubusercontent.com/85239124/197861094-3d97b6fd-aaba-47c3-a9b6-1c2be3e8f64b.png">

I tried removing the problem line, but then I get a different error: 

<img width="1343" alt="Screen Shot 2022-10-25 at 3 20 23 PM" src="https://user-images.githubusercontent.com/85239124/197863000-1bab2b1b-f36e-4802-8e6d-f11f8bcda30a.png">

So, I looked into what path the similarity jar location gave me: 
'd:\\Users\\mmagoffin\\AppData\\Local\\hatch\\env\\virtual\\ds-misc-mmagoffin-S1xDGiIH\\ds-misc-mmagoffin\\lib\\site-packages\\splink\\jars/scala-udf-similarity-0.0.9.jar'

However, looking inside of the splink directory there, the jars directory does not exist. I also tried just copying the Jaro-Winkler jar to the working directory, but I got the same error. 

Any clue how I can fix this?

More context if it helps: I'm on Windows 10, Spark 3.3.0, and splink 3.4.1

And **update**: even though I've been able to fix the secondary jars directory problem, because I'm unable to set a Checkpoint directory here, I'm unable to run linker.estimate_u_using_random_sampling(target_rows=5e5) as I get a checkpointing error within the spark_linker script. Its even more confusing as setCheckpointDir is successful in creating a directory and writing to it, so I'm sure where the error is.

**Comment by mamonu** [comment, marked as answer]
Can you either 
- put the jar in the path :
`d:\Users\mmagoffin\AppData\Local\hatch\env\virtual\ds-misc-mmagoffin-S1xDGiIH\ds-misc-mmagoffin\lib\site-packages\splink\jars/`

- or change the path (where currently you store the output of the function `similarity_jar_location`)
 to your working directory.
 
Also a warning: when you define settings for the spark session you need to restart your kernel in 
order for these to be updated. If you just change the path and run the cell you dont get the jars loaded into the session

**Reply by mmagoffin-sd** [reply]
That worked for the second error, thank you! Do you have any thoughts on what could still be causing the setCheckpointDir error?

**Reply by RobinL** [reply]
I think the `setCheckpointDir` is to do with the difference between the demo setup (in Spark local mode) the Spark setup you're using.  I think you probably need to set the path to a temporary directory accessible to your cluster - probably somewhere in your hadoop file system.

**Reply by mmagoffin-sd** [reply]
I'm currently starting a SparkSession with configuring the master to  "local[4]"- should that be needing a file path to something in my Hadoop file system?

**Reply by RobinL** [reply]
It just needs to be any path to which Spark can write temporary files, wherever that would be in your setup

**Reply by mmagoffin-sd** [reply]
hmm must be an issue with my spark set up then, thank you!

**Reply by mmagoffin-sd** [reply]
update: to be able to use `setCheckpointDir` I had to download the x64 C++ file: https://www.microsoft.com/en-us/download/details.aspx?id=26999 and run the command `[your hadoop dir]\bin\winutils.exe chmod -R 777 [your path to your tmp dir]` in whatever your working directory is
